version 4/03 environments/04 Production Configuration/README.md

```diff
@@ -0,0 +1,341 @@
+    new HtmlWebpackPlugin({
+      filename: 'index.html', //Name of file in ./dist/
+      template: 'index.html', //Name of template in ./src
+      hash: true,
+    }),
+    new MiniCssExtractPlugin({
+      filename: "[name].css",
+      chunkFilename: "[id].css"
+    }),
+  ],
+};
+```
+
+- Now it's time to install `webpack-merge` package. This allow us to combine `base.webpack.config` with environment config:
+
+```
+npm install webpack-merge --save-dev
+```
+
+- Let's go to create the `dev` environment configuration:
+
+### ./dev.webpack.config.js
+```javascript
+const merge = require('webpack-merge');
+const common = require('./base.webpack.config.js');
+
+module.exports = merge(common, {
+    mode: 'development',
+    devtool: 'inline-source-map',
+    devServer: {
+        contentBase: './dist'
+    }
+});
+```
+
+- Make a configuration to see in console the enviroment in which we are
+
+### ./averageService.js
+```diff
+export function getAvg(scores) {
+  return getTotalScore(scores) / scores.length;
+}
+
+export function getTotalScore(scores) {
+  return scores.reduce((score, count) => score + count);
+}
+
++ console.log(`We are in: ${process.env.NODE_ENV}`);
+```
+
+- Finally we need to update command script:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+-   "start": "webpack-dev-server --mode development --open --hot",
++    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+    "build": "rimraf dist && webpack --mode development"
+  },
+  ...
+}
+```
+
+- Running `npm start`:
+
+We can see in console `We are in: development`
+
+- We are going to create a `build dev` command to see how much size has bundles files without `webpack-dev-server` stuff:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+-    "build": "rimraf dist && webpack --mode development"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js"
+  },
+  },
+  ...
+}
+
+```
+
+- Running `npm run build`:
+
+The `dist` folder is created. If we execute `index.html`, it shows us `We are in: development`.
+
+- Let's configure `production` environment:
+
+### ./prod.webpack.config.js
+```javascript
+const merge = require('webpack-merge');
+const common = require('./base.webpack.config.js');
+
+module.exports = merge(common, {
+    mode: 'production',
+});
+
+```
+
+- Add production command script:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+-    "build": "rimraf dist && webpack --config dev.webpack.config.js"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js",
++    "build:prod": "rimraf dist && webpack --config prod.webpack.config.js"
+  },
+  ...
+}
+
+```
+
+- Running `npm run build:prod`, we can see in console `We are in: production` and we can see how all bundle sizes decrease:
```

Comment:
If we runn `npm run build:prod`, the following output will be dumped into the console log `We are in: production` and we can check that all bundle sizes decrease:

Reply:
done